### DISCUSSION TITLE
[RFC] 018 - 多模型服务商一期：架构设计 & AWS Bedrock / Zhipu / Gemini / Moonshot 支持

### DISCUSSION THREAD

**Discussion by arvinxx** [lobehub/lobehub#737] [RFCs]
# 背景

随着 LobeChat 的发展，社区同学对模型服务商的多样性也产生了新的诉求。我们不能只锚着 OpenAI ，而是要丰富模型服务商的多样性，为用户提供更多的会话选择。

1. 支持不同服务商：
- #151 
- [x] AWS Bedrock https://sdk.vercel.ai/docs/guides/providers/aws-bedrock
- [x] Google Model  （Gemini 、Gemini Pro） https://sdk.vercel.ai/docs/guides/providers/google
  - #694 
  - #656 
  - #650 
  - #213 
  - #798 
  - #1046
- [x] ChatGLM #1125
- [x] Moonshot AI 月之暗面 https://github.com/lobehub/lobe-chat/pull/1232
- [ ] replicate https://sdk.vercel.ai/docs/guides/providers/replicate
- [ ] MinMax #1053  

2. 支持本地模型 (详见 #1283)
3. 支持同一模型的不同 API Key (详见 #1287)
4. 支持更完整的 Azure OpenAI 配置 （详见  #1219）
5. 其他 hack 的方案
- #1093

**Comment by arvinxx** [comment]
## 技术选型讨论

AI SDK VS LangChain

大的方向上，我觉得 chatCompletion 这个模块仍然基于 OpenAI 和 Vercel AI SDK ，不用 LangChain 了，流式输出 functionCall 的问题已经过去快一周了，官方都没解决。

另外，现在整个社区的趋势就是接口调用的风格和规范会尽量对齐 OpenAI 的 chat completion。因此用 openai + vercel ai sdk 基本上可以满足 80% 主流模型的诉求 。

> 回复存放之前的一些探索方案

**Reply by arvinxx** [reply]
## 对话重构为 LangChain 实现


### 坑点 1：LangChain 官方实现对 stream 中的错误处理不完善

参考[官方文档](https://js.langchain.com/docs/modules/model_io/chat/streaming)，可以使用 `pipe` + `parser` 来简化 stream 实现：

```ts
 try {
    const parser = new HttpResponseOutputParser();
    const stream = await model.pipe(parser).stream(messages);

    return new Response(stream);
  } catch (error) {
  // error handle
}

```
如果请求是正常的，那么上述代码可以正常运行。

但一个巨坑的地方是如果stream 出错，catch 不会捕获错误，会直接退出运行。

已提 issue : https://github.com/langchain-ai/langchainjs/issues/4015 。



### 坑点2： `HttpResponseOutputParser` 无法正常输出 OpenAI Function

HttpResponseOutputParser 类无法正常解析 tool_calls 的输出，导致返回的结果是空的。

![image](https://github.com/lobehub/lobe-chat/assets/28616219/fed7eb24-374f-420c-8b83-5561b7644fba)

issue： https://github.com/langchain-ai/langchainjs/issues/4095

**Comment by arvinxx** [comment]
## 错误处理记录

### Zhipu

初始化

- 密码不正确 InvalidAccessCode
- API 出错 InvalidZhipuAPIKey


Chat 运行态


`ZhipuBizError`: Zhipu业务逻辑错误（参数非法、token 不正确等） 
```
{
  "code": "1002",
  "message": "Authorization Token非法，请确认Authorization Token正确传递。"
}
```

`AgentRuntimeError`: Zhipu 模块运行出错（Lobe Runtime 侧的代码实现有问题触发报错）
```json
{
  "message": "test",
  "name": "TypeError",
  "stack": "TypeError: test\n    at LobeZhipuAI.chat (webpack-internal:///(rsc)/./src/libs/agent-runtime/zhipu/index.ts:53:19)\n"
}
```

**Comment by snowyu** [comment]
@arvinxx The error handling with stream should be:

```ts
stream.on('error', (e) => handleError(e))
.pipe(b)
.on('error', (e) => handleError(e))
.pipe(c)
.on('error', (e) => handleError(e));

function handleError(e) { throw e; } // Maybe also some logging or smth.
```

**Reply by arvinxx** [reply]
thanks, I will try~

**Reply by arvinxx** [reply]
@snowyu it doesn't work. But LangChain have solved this problem officially.

**Comment by arvinxx** [comment]
## 节奏

分为三期进行：

- 一期：#1173
  - 支持业务诉求和用户呼声较多的三个服务商；
  - 初步搭建第一版多模型服务商下的数据管理模型（前端）；
  - 抽出初步的 Agent Runtime 调用运行时；
 
- 二期：
  - 其余模型服务商；
  - 本地 LLM 支持；

- 三期：
  - Agent Runtime #924 
  - 完善服务商配置

**Comment by arvinxx** [comment]
## Logo 

Gemini https://icon-sets.iconify.design/logos/google-gemini/

**Reply by canisminor1990** [reply]
约等于 google bard, gemini 的 logo 直接做 icon感觉太大了

Google Bard: https://simpleicons.org/?q=google+bard
Meta Llama: https://simpleicons.org/?q=meta
OpenAI: https://simpleicons.org/?q=OpenAI

ChatGLM 的他们还没传 simpleicons 么

**Reply by canisminor1990** [reply]
想搞个 llm logo & icon 仓库了

**Reply by arvinxx** [reply]
要分两种

一种是model tag，我初步搞好了，在 https://github.com/lobehub/lobe-chat/blob/feat/model-provider/src/components/ModelTag/index.tsx

另外一种在设置里配置模型服务商，倒是可以用这种大的，更容易凸显品牌

**Comment by arvinxx** [comment]
## 实现思路

### 整体思路

- [x] 1. model层调通
- [X] 2. 路由改造，支持多模型服务商路由
- [x] 3. 前端界面设置，支持配置服务商 api。目前由于settings部分需要globalStore进行水合，导致无法实现服务端数据的准确merge。需要先行改造 #1180
- [x] 4. Chat会话界面改造 ModelTag，ModelSelect


### LobeAIProvider

支持接口
1. 初始化
2. ChatCompletition

错误类型
1. Apikey 错误
2. 欠费
3. 超过token数
4. RateLimitted

**Reply by arvinxx** [reply]
本次架构升级的考虑要点

## 模型元信息维护

需要有一个全局视角来维护与管理模型信息，包括 model id ，展示名称，maxToken ，是否支持function call 与 vision 等。

有了这个信息以后，我们就可以有一个更好的全局视角来管理模型

## 错误处理框架

和单一的模型服务商不再一样，需要精细化梳理各个模型服务商提供的错误类型，然后找到公共部分与差异化部分，并用一个统一的错误处理方案来承接。

## 服务端配置、客户端配置的合并与可视化配置

目前已经实现了从服务端读取配置作为客户端的默认值，并且支持用户在前端界面进行个性化配置

这里后续需要重点的是服务端可控性，提供一种灵活的方式允许服务端是否开启或关闭

## OneAPI 兼容性

目前存在大量的用户仍在使用 oneapi 的方案来统一管理模型，我们需要提供一种机制来让这部分用户仍然可以进行相应的配置与管理。

**Comment by arvinxx** [comment]
需要注意的点：
1. 不同的APIKey错误要唤起不同的api填写表单
2. 后续需要优化 agent runtime 的实例初始化方案
3. ModelList 的自定义方案

**Comment by snowyu** [comment]
对于配置表单UI建议用JSONSchema统一生成: https://github.com/rjsf-team/react-jsonschema-form/

**Reply by arvinxx** [reply]
感谢建议，后续可以试试看

**Comment by arvinxx** [comment]
## Zhipu AI 接入踩坑点记录

由于智谱 V4 的API 朝 OpenAI 对齐了，所以整体接入比较顺利，不过仍然存在几个坑点：

### 图片base64需要移除前缀；

智谱不支持 `data:image/png;base64,UklGRqTNAABXRUJQVlA4WAoAAAA` ，需要去掉前面的 `data:image/png;base64,` 

### `tool_calls` 流式有问题：

标准 OpenAI 的 tool_calls 输出如下：
```jsonl
chunk 0
{"choices":[{"index":0,"delta":{"role":"assistant","content":null,"tool_calls":[{"index":0,"id":"call_EOjIHZ8INcmtndup397Iy2R0","type":"function","function":{"name":"realtime-weather____fetchCurrentWeather","arguments":""}}]},"logprobs":null,"finish_reason":null}]}

chunk 1
{"choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":"{\""}}]},"logprobs":null,"finish_reason":null}]}

chunk 2
{"choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":"city"}}]},"logprobs":null,"finish_reason":null}]}

chunk 3
{"choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":"\":\""}}]},"logprobs":null,"finish_reason":null}]}

chunk 4
{"choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":"Hang"}}]},"logprobs":null,"finish_reason":null}]}

chunk 5
{"choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":"zhou"}}]},"logprobs":null,"finish_reason":null}]}

chunk 6
{"choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"arguments":"\"}"}}]},"logprobs":null,"finish_reason":null}]}

chunk 7
{"choices":[{"index":0,"delta":{},"logprobs":null,"finish_reason":"tool_calls"}]}
```

但 zhipu 的接口输出格式如下：

```jsonl
chunk 0
{"choices":[{"index":0,"finish_reason":"tool_calls","delta":{"role":"assistant","tool_calls":[{"id":"call_8349638559205414584","index":0,"type":"function","function":{"name":"realtime-weather____fetchCurrentWeather","arguments":"{\"city\":\"杭州\"}"}}]}}]}

chunk 1
{"choices":[{"index":0,"finish_reason":"tool_calls","delta":{"role":"assistant","content":""}}],"usage":{"prompt_tokens":153,"completion_tokens":20,"total_tokens":173}}
```

可以看到最后的一个输出的 chunk 是，delta 是 `{"role":"assistant","content":""}`，这个行为和 openai 的不一致。这就导致最终前端收到的 arguments 参数为空字符串，所以会出现问题

**Reply by arvinxx** [reply]
一期结论：针对 GLM 系列模型，本次上线版本先不放出 FC 功能，等智谱侧修复后再上线


Vision Model 由于集成 Google Genimini 时做了一下 base64 的解析与转换，因此可以复用实现兼容，等到时候同样更新修复后移除

**Reply by MapleEve** [reply]
~~在 `action.ts` 中打个补丁会不会更好一点，因为GLM4的长文件处理效果比 GPT4 差不多，价格又便宜很多。~~

放弃了，等他官方支持再做流式

**Comment by arvinxx** [comment]
## AWS Bedrock 接入笔记

基本上参考 https://sdk.vercel.ai/docs/guides/providers/aws-bedrock 来接入会比较顺利。

初始化client:
```ts
const bedrockClient = new BedrockRuntimeClient({
  region: process.env.AWS_REGION ?? 'us-east-1',
  credentials: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID ?? '',
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY ?? '',
  },
});
```

chat 接口：
```ts
  // Ask Claude for a streaming chat completion given the prompt
  const bedrockResponse = await bedrockClient.send(
    new InvokeModelWithResponseStreamCommand({
      modelId: 'anthropic.claude-v2',
      contentType: 'application/json',
      accept: 'application/json',
      body: JSON.stringify({
        prompt: experimental_buildAnthropicPrompt(messages),
        max_tokens_to_sample: 300,
      }),
    }),
  );
 
  // Convert the response into a friendly text-stream
  const stream = AWSBedrockAnthropicStream(bedrockResponse);
 
  // Respond with the stream
  return new StreamingTextResponse(stream);
```


其中 `max_tokens_to_sample` 是必填项，否则会报错。另外比较有坑的是模型的 id，一开始配置的是这几个：
```
anthropic.claude-instant-v1
anthropic.claude-v2:1
meta.llama2-13b-chat-v1
meta.llama2-70b-chat-v1
```

但在接入 llama2 时则遇到了问题， llama 的 prompt 结构和 aws 的会有很大差别，比如 `max_tokens_to_sample` 就不支持，会报错：
```js
{
  body: {
  httpStatusCode: 400,
  requestId: '*',
  extendedRequestId: undefined,
  cfId: undefined,
  attempts: 1,
  totalRetryDelay: 0
},
  message: 'Malformed input request: #: extraneous key [max_tokens_to_sample] is not permitted, please reformat your input and try again.',
  region: 'us-east-1',
  type: 'ValidationException'
}
```

所以计划一期先不上 llama2。

**Reply by arvinxx** [reply]
更新：Vercel AI SDK 有 llama 的示例： https://sdk.vercel.ai/docs/api-reference/aws-bedrock-stream#example-llama-2-models 。试了下可以用，因此本期可以上。

**Comment by arvinxx** [comment]
## Google Gemini 接入

核心是代理问题，所以参考 https://github.com/google/generative-ai-js/issues/29#issuecomment-1866246513 在 provider route 加了一个 proxy，在 proxy 模式下，使用 nodejs runtime，线上仍然使用 edge runtime

```ts

// due to the Chinese region does not support accessing Google / OpenAI
// we need to use proxy to access it
const proxyUrl = process.env.HTTP_PROXY_URL;
const useProxy = !!proxyUrl;

if (useProxy) {
  const { setGlobalDispatcher, ProxyAgent } = require('undici');

  setGlobalDispatcher(new ProxyAgent({ uri: proxyUrl }));
}
// undici only can be used in NodeJS.
// So when using proxy, switch to NodeJS runtime
export const runtime = useProxy ? 'nodejs' : 'edge';
```

**Comment by arvinxx** [comment]
# 后续如何集成一个新 Provider




### modelProviders 元信息

`config/modelProviders` 新建一个文件，填写模型服务商基础信息和模型基础信息

```ts
import { ModelProviderCard } from '@/types/llm';

const ZhiPu: ModelProviderCard = {
  chatModels: [
    {
      description: '最新的 GLM-4 、最大支持 128k 上下文、支持 Function Call 、Retreival',
      displayName: 'GLM-4',
      // functionCall: true,
      id: 'glm-4',
      tokens: 128_000,
    },
    {
      description:
        '实现了视觉语言特征的深度融合，支持视觉问答、图像字幕、视觉定位、复杂目标检测等各类多模态理解任务',
      displayName: 'GLM-4 Vision',
      id: 'glm-4v',
      tokens: 128_000,
      vision: true,
    },
    {
      description: '最新的glm-3-turbo、最大支持 128k上下文、支持Function Call、Retreival',
      displayName: 'GLM-3 Turbo',
      // functionCall: true,
      id: 'glm-3-turbo',
      tokens: 128_000,
    },
  ],
  id: 'zhipu',
};

export default ZhiPu;
```

## 服务端实现

### 添加环境变量

在 `src/config/server/provider.ts` 添加新的环境变量

### 新增runtime

在 runtime 新增实现 LobeRuntimeAI，

需要包包含初始化、 chat 接口，并需要实现初始化错误、 业务逻辑错误这两个错误处理

### 路由集成

在 chat/provider/route 接入实现

### 默认配置集成

在 api/config 设定发给客户端是否开启的属性

## 客户端实现

### 模型/服务商 icon

Components/ModelIcon 补充判断逻辑

### selectors

添加是否启用的逻辑

### settings 

增加一个设置面板渲染组件

### api配置面板

在error/apikeyform新增配置表单



------


服务端：
1. 服务端环境变量 ：`config/server/provider.ts` 新增 MOONSHOT_API_KEY 等环境变量；
2. Moonshot 运行时 `libs/agent-runtime` 新增 LobeMoonshotAI ，
  - 实现 `init` 和 `chat` 两个方法；
  - 完善相应的  `InvalidMoonshotAPIKey` 和 `MoonshotBizError` 两个错误状态
3. 路由定义： `api/chat/[provider]/agentRuntime.ts` 新增 moonshotRuntime 初始化逻辑（客户端key 与 环境变量）
4. 服务端配置: `api/config/route.ts` 定义 Provider 默认开启逻辑
5. 错误响应状态码处理： `api/errorResponse.ts` 定义 `MoonshotBizError` 状态码

客户端：
1. 模型元数据定义（`config/modelProviders`），补充完善相应的模型元信息
2. 模型设置： `types/settings/modelProvider` 定义 `MoonshotConfig`，在 `const/settings` 设定默认值；
3. 取数 seletors： `store/global/slices/settings/selector/modelProvider.ts` 补充取数 selector （enableMoonshot、 moonshotAPIKey）
4. 前端鉴权 token： `services/_auth` 补充获取 apiKey 相关逻辑；
5. 模型UI （模型选择器、模型 Tag ）：
   - Provider Logo: `components/ModelProviderIcon` 更新
   - Model Logo: `components/ModelIcon` 更新
   - Model Tag: `components/ModelTag` 更新
6. 错误处理 `features/Conversation/Error` 新增一种对应的 APIKeyForm 
7. 全局模型设置 `app/settings/llm` 新增一个配置组件（可以拷贝复用 Zhipu ）

**Reply by sjy** [reply]
Really helpful, but I can't find the flow to update locales p.s: I don't have a valid OPEN_AI_KEY yet

**Reply by gijigae** [reply]
@arvinxx, just want to say "Thank you!" not only for all the changes you have maintained, but also for the knowledge you have shared. The steps to add Moonshot really helped me understand the overall workflow around onboarding a new model provider. Keep up the good work!

**Comment by arvinxx** [comment]
## 特性跟踪


-----

## Google Gemini 支持

<img width="1317" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/35e60f72-8d27-4259-a010-45d3b6372837">

<img width="679" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/dfd54872-8b75-4cb0-a6c6-8346c9b80636">

但个人感觉 Gemini Pro Vision 的视觉能力非常垃圾，还不支持多轮对话，根本没法在日常中使用

## AWS Bedrock 支持

- 支持 AWS Bedrock （Claude / LLAMA 2）

<img width="1058" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/6e6b7a7d-905b-413c-8659-a73e062a3344">

## 智谱 GLM 支持

- 支持智谱 GLM 系列

<img width="1048" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/d3aa36ab-084c-41ab-86c2-752dc411581b">

个人试用下来真的比 Gemini 、Bedrock 之类的好太多了！！！ 国货之光！！！


## Azure OpenAI 接入重构

TODO

**Reply by arvinxx** [reply]
## API Key 填写

| OpenAI | Zhipu | Google | Bedrock | 
|---|---|---|---|
|<img width="586" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/28bab5fa-77a6-4a04-af8f-07bd40b2f0df">|<img width="510" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/2f812665-d96e-4fc6-878d-904e9e2d6a52">|<img width="502" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/b78c5b4c-5360-41b7-9902-da224572ea4a">|<img width="505" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/5459cb01-eed3-4b61-b05f-c601f2558dcd">|
|<img width="510" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/e8d91596-cc2a-45d8-9999-3c31eb589649">|<img width="514" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/df561bfc-c69c-4170-9a38-b40c531bfe5d">|<img width="519" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/1afe7970-80a6-4a5c-9583-1ce7cef17a8a">|<img width="505" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/4eeb59d3-a9d8-4f03-a421-e994b2deffa2">|


对比


| OpenAI | Zhipu | Google | Bedrock | 
|---|---|---|---|
|<img width="509" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/2eaca5ca-eae5-4091-91cd-f621b607e325">|<img width="506" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/d4e5d6a6-2661-436b-834d-c0e195aa54a7">|<img width="517" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/622146bf-06f7-4028-80a3-8fd5ef9e8eda">|<img width="611" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/af0cb88f-38e7-4c2e-ac5b-0c98577ff3ac">|
|<img width="581" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/e3329d9c-db90-4414-9c6c-aecf53903438">|<img width="520" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/f3a33564-a97d-4af9-963c-7707626e9bc4">|<img width="519" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/1afe7970-80a6-4a5c-9583-1ce7cef17a8a">|<img width="509" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/420f216c-9997-4fe0-9122-6b5d8d77f32b">|
